Gh 7202 mvcc refactoring

alyapunov opened the pull request:
A series of refactoring commits in MVCC. A couple of bugs were fixed and options in box.atomic are introduced.

Closes #7945 and #7202.

CuriousGeorgiy commented on src/box/memtx_space.c:
```diff
@@ -276,7 +276,7 @@ memtx_space_replace_all_keys(struct space *space, struct tuple *old_tuple,
 	 * Don't use MVCC engine for ephemeral in any case.
```
Nit: I refer to this subsystem as `memtx`, not `txm` (just to be on the same page).

drewdzzz commented on src/box/memtx_tx.c:
```diff
@@ -2095,8 +2095,7 @@ memtx_tx_handle_gap_write(struct txn *txn, struct space *space,
 		if (fsc_item->txn != txn &&
-		    (memtx_tx_cause_conflict(txn, fsc_item->txn) != 0 ||
-		     memtx_tx_track_read_story(fsc_item->txn, space, story,
+		    (memtx_tx_track_read_story(fsc_item->txn, space, story,
```
It's not an obvious place. We should cover it with comment. Also, all similar places should be covered.

CuriousGeorgiy replied:
Discussed with @drewdzzz and @alyapunov: perhaps rename the function to `memtx_tx_story_link_read_by` (similarly to `memtx_tx_story_link_{added, deleted}_by}`) and add a concise description to this function (currently, it does not have any).

CuriousGeorgiy commented on test/box-luatest/gh_7202_excess_conflict_secondary_test.lua:
```diff
@@ -0,0 +1,101 @@
+            {{1, 'test - test', 'READY', 'test'},
+             {2, 'test - test', 'READY', 'test'}})
+    end)
+end
```
TBH, I had a really hard time udnerstanding what's going on this test. Also, it seems like this test relies on the order of fiber execution. AFAIC, `secondUpdate` can actually be scheduled before `firstUpdate`: I left only one `put` fiber and observed exactly this.

I suggest you also add this simplified version (will come back a little bit later with it):
```lua

```

CuriousGeorgiy commented on test/box-luatest/gh_7202_options_in_atomic_test.lua:
```diff
@@ -0,0 +1,202 @@
+            {{1, 'atomic1', 'function'},
+             {2, 'atomic2', 'table'}})
+    end)
+end
```
I thought we had a function to get the current transaction's isolation level, but unfortunately we do not have one. If we did — I would rather test that the isolation level is correct.

TBH, it's kind of hard for to understand what's going on these tests: as I understand, you are basically testing that the isolation levels work correctly — but we already have [gh-6390](https://github.com/tarantool/tarantool/blob/master/test/engine-luatest/gh_6930_mvcc_isolation_levels_test.lua) test for this.

CuriousGeorgiy commented on src/box/memtx_tx.c:
```diff
@@ -1780,9 +1791,16 @@ memtx_tx_story_find_visible_tuple(struct memtx_story *story, struct txn *tnx,
 		assert(index < story->index_count);
-		if (memtx_tx_story_is_visible(story, tnx, is_prepared_ok,
-					      visible_tuple, is_own_change))
+		if (memtx_tx_story_end_is_visible(story, tnx, is_prepared_ok,
```
We have the notion of a story beginning (the comment could be reworded to `The statement that began this story.`:
https://github.com/tarantool/tarantool/blob/182034dc979b3125f0ad66227997b58b400b6e2e/src/box/memtx_tx.h#L158-L164

We also have the the notion of a story ending:
https://github.com/tarantool/tarantool/blob/182034dc979b3125f0ad66227997b58b400b6e2e/src/box/memtx_tx.h#L171-L176

I am okay with the current naming, perhaps we could change it to `story_{begininng, ending}` or stick with the current names for brevity.

drewdzzz replied:
The presence of these terms does not contradict the fact that my version is more illustrative. In the context of the places where these functions are used, we are checking to see if the transaction sees the tuple deletion and addition. So why increase the cognitive complexity of the code and use terms that can be found somewhere in the description to a field that even has a different name?

CuriousGeorgiy commented on src/box/memtx_tx.c:
```diff
@@ -489,7 +469,6 @@ memtx_tx_region_alloc(struct txn *txn, size_t size,
 /** String representation of enum memtx_tx_alloc_type. */
 const char *memtx_tx_alloc_type_strs[MEMTX_TX_ALLOC_TYPE_MAX] = {
 	"trackers",
-	"conflicts",
```
Perhaps we should consider increasing the granulrity of allocation types, i.e., by adding specific tracker types, e.g.: "point hole", "gap", "full scan", "read", "conflict".

drewdzzz replied:
I agree that the monitoring is not very informative now. But it is made mostly for users, not for developers. This kind of detail looks useless - such detailed metrics simply do not give any information to the user - and he is likely to have many questions about what the "point hole" is.

alyapunov commented on src/box/memtx_tx.c:
```diff
@@ -1885,193 +1836,80 @@ check_hole(struct space *space, uint32_t index,
  * Check that replaced tuples in space's indexes does not violate common
  * replace rules. See memtx_space_replace_all_keys comment.
- * (!) Version for the case when replaced tuple is dirty.
+ * Call check_dup_clean or check_dup_dirty depending on situation.
```
Thanks, fixed.

alyapunov commented on test/box-luatest/gh_7945_mvcc_assert_in_mvcc_test.lua:
```diff
@@ -0,0 +1,125 @@
+        box.space.test:create_index('primary', {type = 'tree'})
+
+        local t = require('luatest')
+        local txn_proxy = require('test.box.lua.txn_proxy')
```
I like txn_proxy more because it is much simpler.. For example I has my own version of txn_proxy which emits lots of debug information, that allows me to debug failing tests.

alyapunov commented:
I fixed most all places you commented and replied the rest. Please check.
Also I made a significant changes:
1. Created an issue #8122 and rebound most of the commits to it. I made it because actually #7202 is about a feature, while I fixed a bug that was needed to be fixed prior to that feature. Also I want to push all the commits except the last to 2.10 as bugfixes.
2. Reverted removal of monitoring metrics.
3. Introduces `box.internal.txn_isolation()` which allowed to simplify significantly the test in the last commit.

I saved the old branch and you can simply check my after-review fixes by the command:
```
c=1 && git show alyapunov/gh-7202-mvcc-refactoring-before-review~$c > 1.txt && git show alyapunov/gh-7202-mvcc-refactoring~$c > 2.txt && meld 1.txt 2.txt
```
use c=0..9, and perhaps you need to use `diff` instead of `meld`.

CuriousGeorgiy commented on test/box-luatest/gh_7945_mvcc_assert_in_mvcc_test.lua:
```diff
@@ -0,0 +1,125 @@
+        tx2:begin()
+
+        -- Set tx1 to track interval from minus infinity to {10}.
+        tx1("box.space.test:select({}, {limit=1})")
```
I meant that you could parametrize your test with the index type, and do `tx1("box.space.test:select({})` here.

alyapunov commented:
Issue about metrics: https://github.com/tarantool/tarantool/issues/8149

alyapunov commented:
Backported to 2.10 all except the last commit (which is feature).
